# <font color='red'><b>Bootstrap assignment</b> </font>

<b>There will be some functions that start with the word "grader" ex: grader_sampples(), grader_30().. etc, you should not change those function definition.

Every Grader function has to return True.</b>

<font color='blue'> <b>Importing packages</b> </font>

In [112]:
import numpy as np # importing numpy for numerical computation
from sklearn.datasets import load_boston # here we are using sklearn's boston dataset
from sklearn.metrics import mean_squared_error # importing mean_squared_error metric
from sklearn.tree import DecisionTreeRegressor
from sklearn.metrics import mean_squared_error
import random
import pickle

In [2]:
boston = load_boston()
x=boston.data #independent variables
y=boston.target #target variable

/usr/local/lib/python3.7/dist-packages/sklearn/utils/deprecation.py:87: FutureWarning: Function load_boston is deprecated; `load_boston` is deprecated in 1.0 and will be removed in 1.2.

    The Boston housing prices dataset has an ethical problem. You can refer to
    the documentation of this function for further details.

    The scikit-learn maintainers therefore strongly discourage the use of this
    dataset unless the purpose of the code is to study and educate about
    ethical issues in data science and machine learning.

    In this special case, you can fetch the dataset from the original
    source::

        import pandas as pd
        import numpy as np


        data_url = "http://lib.stat.cmu.edu/datasets/boston"
        raw_df = pd.read_csv(data_url, sep="\s+", skiprows=22, header=None)
        data = np.hstack([raw_df.values[::2, :], raw_df.values[1::2, :2]])
        target = raw_df.values[1::2, 2]

    Alternative datasets include the California housing dataset (i.e.

In [3]:
x.shape

(506, 13)

In [4]:
x[:5]

array([[6.3200e-03, 1.8000e+01, 2.3100e+00, 0.0000e+00, 5.3800e-01,
        6.5750e+00, 6.5200e+01, 4.0900e+00, 1.0000e+00, 2.9600e+02,
        1.5300e+01, 3.9690e+02, 4.9800e+00],
       [2.7310e-02, 0.0000e+00, 7.0700e+00, 0.0000e+00, 4.6900e-01,
        6.4210e+00, 7.8900e+01, 4.9671e+00, 2.0000e+00, 2.4200e+02,
        1.7800e+01, 3.9690e+02, 9.1400e+00],
       [2.7290e-02, 0.0000e+00, 7.0700e+00, 0.0000e+00, 4.6900e-01,
        7.1850e+00, 6.1100e+01, 4.9671e+00, 2.0000e+00, 2.4200e+02,
        1.7800e+01, 3.9283e+02, 4.0300e+00],
       [3.2370e-02, 0.0000e+00, 2.1800e+00, 0.0000e+00, 4.5800e-01,
        6.9980e+00, 4.5800e+01, 6.0622e+00, 3.0000e+00, 2.2200e+02,
        1.8700e+01, 3.9463e+02, 2.9400e+00],
       [6.9050e-02, 0.0000e+00, 2.1800e+00, 0.0000e+00, 4.5800e-01,
        7.1470e+00, 5.4200e+01, 6.0622e+00, 3.0000e+00, 2.2200e+02,
        1.8700e+01, 3.9690e+02, 5.3300e+00]])

## <font color='red'><b>Task 1</b></font>

<font color='red'> <b>Step - 1</b></font>

*  <font color='blue'><b>Creating samples</b></font><br>
    <b> Randomly create 30 samples from the whole boston data points</b>
    *  Creating each sample: Consider any random 303(60% of 506) data points from whole data set and then replicate any 203 points from the sampled points
    
     For better understanding of this procedure lets check this examples, assume we have 10 data points [1,2,3,4,5,6,7,8,9,10], first we take 6 data points randomly , consider we have selected [4, 5, 7, 8, 9, 3] now we will replicate 4 points from [4, 5, 7, 8, 9, 3], consder they are [5, 8, 3,7] so our final sample will be [4, 5, 7, 8, 9, 3, 5, 8, 3,7]
* <font color='blue'><b> Create 30 samples </b></font>
    *  Note that as a part of the Bagging when you are taking the random samples <b>make sure each of the sample will have different set of columns</b><br>
Ex: Assume we have 10 columns[1 ,2 ,3 ,4 ,5 ,6 ,7 ,8 ,9 ,10] for the first sample we will select [3, 4, 5, 9, 1, 2] and for the second sample  [7, 9, 1, 4, 5, 6, 2] and so on...
Make sure each sample will have atleast 3 feautres/columns/attributes

* <font color='red'><b> Note - While selecting the random 60% datapoints from the whole data, make sure that the selected datapoints are all exclusive, repetition is not allowed. </b></font>

<font color='red'><b>Step - 2 </b></font>

<font color='blue'><b>Building High Variance Models on each of the sample and finding train MSE value</b></font>

*  Build a regression trees on each of 30 samples.
*  Computed the predicted values of each data point(506 data points) in your corpus.
*  Predicted house price of $i^{th}$ data point $y^{i}_{pred} =  \frac{1}{30}\sum_{k=1}^{30}(\text{predicted value of } x^{i} \text{ with } k^{th} \text{ model})$
*  Now calculate the $MSE =  \frac{1}{506}\sum_{i=1}^{506}(y^{i} - y^{i}_{pred})^{2}$

<font color='red'> <b>Step - 3 </b></font>

*  <font color='blue'><b>Calculating the OOB score </b></font>

*  Predicted house price of $i^{th}$ data point $y^{i}_{pred} =  \frac{1}{k}\sum_{\text{k= model which was buit on samples not included } x^{i}}(\text{predicted value of } x^{i} \text{ with } k^{th} \text{ model})$.
*  Now calculate the $OOB Score =  \frac{1}{506}\sum_{i=1}^{506}(y^{i} - y^{i}_{pred})^{2}$.

# <font color='red'><b>Task 2</b></font>

*  <font color='blue'><b>Computing CI of OOB Score and Train MSE</b></font>
  *   Repeat Task 1 for 35 times, and for each iteration store the Train MSE and OOB score </li>
<li> After this we will have 35 Train MSE values and 35 OOB scores </li>
<li> using these 35 values (assume like a sample) find the confidence intravels of MSE and OOB Score </li>
<li> you need to report CI of MSE and CI of OOB Score </li>
<li> Note: Refer the Central_Limit_theorem.ipynb to check how to find the confidence intravel</li>
</ol>

# <font color='red'><b>Task 3</b></font>

*  <font color='blue'><b>Given a single query point predict the price of house.</b></font>

Consider xq= [0.18,20.0,5.00,0.0,0.421,5.60,72.2,7.95,7.0,30.0,19.1,372.13,18.60] 
Predict the house price for this point as mentioned in the step 2 of Task 1.

## <font color='red'><b>A few key points</b></font>

* Remember that the datapoints used for calculating MSE score contain some datapoints that were initially used while training the base learners (the 60% sampling). This makes these datapoints partially seen (i.e. the datapoints used for calculating the MSE score are a mixture of seen and unseen data).
Whereas, the datapoints used for calculating OOB score have only the unseen data. This makes these datapoints completely unseen and therefore appropriate for testing the model's performance on unseen data.

* Given the information above, if your logic is correct, the calculated MSE score should be less than the OOB score.

* The MSE score must lie between 0 and 10.
* The OOB score must lie between 10 and 35.

* The difference between the left nad right confidence-interval values must not be more than 10. Make sure this is true for both MSE and OOB confidence-interval values.

# <font color='blue'> <b>Task - 1</b></font>

<font color='blue'><b>Step - 1</b></font>

*  <font color='blue'> <b>Creating samples</b></font>

<font color='Orange'><b>Algorithm</b></font>

![alt text](https://i.imgur.com/OfcFrUP.jpg/)

*  <font color='blue'><b> Write code for generating samples</b></font>

In [99]:
boston = load_boston()
x=boston.data #independent variables
y=boston.target #target variable

/usr/local/lib/python3.7/dist-packages/sklearn/utils/deprecation.py:87: FutureWarning: Function load_boston is deprecated; `load_boston` is deprecated in 1.0 and will be removed in 1.2.

    The Boston housing prices dataset has an ethical problem. You can refer to
    the documentation of this function for further details.

    The scikit-learn maintainers therefore strongly discourage the use of this
    dataset unless the purpose of the code is to study and educate about
    ethical issues in data science and machine learning.

    In this special case, you can fetch the dataset from the original
    source::

        import pandas as pd
        import numpy as np


        data_url = "http://lib.stat.cmu.edu/datasets/boston"
        raw_df = pd.read_csv(data_url, sep="\s+", skiprows=22, header=None)
        data = np.hstack([raw_df.values[::2, :], raw_df.values[1::2, :2]])
        target = raw_df.values[1::2, 2]

    Alternative datasets include the California housing dataset (i.e.

In [10]:
def generating_samples(input_data, target_data):
  selecting_rows = list(np.random.choice(len(x), 303,replace= False))
  #print(selecting_rows)
  #print(len(selecting_rows))

  replacing_rows = list(np.random.choice(selecting_rows,size = 203,replace= True))
  #print(replacing_rows)
  #print(len(replacing_rows))

  rand = random.randint(3,13)
  selecting_columns = list(np.random.choice(13,rand,replace= False))
  #print(selecting_columns)

  sample_data = x[selecting_rows][:,selecting_columns]
  
  target_of_sample_data = y[selecting_rows]

  replicated_sample_data = x[replacing_rows][:,selecting_columns]
  target_of_replicated_sample_data = y[replacing_rows]

  final_sample_data = np.vstack((sample_data,replicated_sample_data))

  final_target_data = np.vstack((target_of_sample_data.reshape(-1,1),target_of_replicated_sample_data.reshape(-1,1)))
  

  '''In this function, we will write code for generating 30 samples '''
    # you can use random.choice to generate random indices without replacement
    # Please have a look at this link https://docs.scipy.org/doc/numpy-1.16.0/reference/generated/numpy.random.choice.html for more details
    # Please follow above pseudo code for generating samples 
    

  return final_sample_data , final_target_data,selecting_rows,selecting_columns
    #note please return as lists'''

<font color='cyan'> <b> Grader function - 1 </b> </fongt>

In [11]:
def grader_samples(a,b,c,d):
    length = (len(a)==506  and len(b)==506)
    sampled = (len(a)-len(set([str(i) for i in a]))==203)
    rows_length = (len(c)==303)
    column_length= (len(d)>=3)
    assert(length and sampled and rows_length and column_length)
    return True
a,b,c,d = generating_samples(x, y)
grader_samples(a,b,c,d)

True

*  <font color='blue'> <b>Create 30 samples </b>

![alt text](https://i.imgur.com/p8eZaWL.jpg)

In [14]:
# Use generating_samples function to create 30 samples 
# store these created samples in a list
list_input_data =[]
list_output_data =[]
list_selected_row= []
list_selected_columns=[]

for i in range(0,30):
  a,b,c,d = generating_samples(x,y)
  list_input_data.append(a)
  list_output_data.append(b)
  list_selected_row.append(c)
  list_selected_columns.append(d)


<font color='cyan'> <b>Grader function - 2 </b></font>

In [15]:
def grader_30(a):
    assert(len(a)==30 and len(a[0])==506)
    return True
grader_30(list_input_data)

True

<br>

<font color='red'><b>Step - 2 </b></font>

<font color='orange'><b>Flowchart for building tree</b></font>

![alt text](https://i.imgur.com/pcXfSmp.png)

*  <font color='blue'><b> Write code for building regression trees</b></font>

In [44]:
#https://machinelearningmastery.com/save-load-machine-learning-models-python-scikit-learn/

list_input_data 
list_output_data 
list_selected_row
list_selected_columns

list_of_all_models = []

for i in range(30):

  model = DecisionTreeRegressor(max_depth=None)
  model.fit(list_input_data[i],list_output_data[i])
  filename = "finalized_model{num}.sav".format(num=i)
  pickle.dump(model, open(filename, 'wb'))
  list_of_all_models.append(filename)


print(len(list_of_all_models))


In [71]:
list_of_all_models

['finalized_model0.sav',
 'finalized_model1.sav',
 'finalized_model2.sav',
 'finalized_model3.sav',
 'finalized_model4.sav',
 'finalized_model5.sav',
 'finalized_model6.sav',
 'finalized_model7.sav',
 'finalized_model8.sav',
 'finalized_model9.sav',
 'finalized_model10.sav',
 'finalized_model11.sav',
 'finalized_model12.sav',
 'finalized_model13.sav',
 'finalized_model14.sav',
 'finalized_model15.sav',
 'finalized_model16.sav',
 'finalized_model17.sav',
 'finalized_model18.sav',
 'finalized_model19.sav',
 'finalized_model20.sav',
 'finalized_model21.sav',
 'finalized_model22.sav',
 'finalized_model23.sav',
 'finalized_model24.sav',
 'finalized_model25.sav',
 'finalized_model26.sav',
 'finalized_model27.sav',
 'finalized_model28.sav',
 'finalized_model29.sav']

In [108]:
all_pred = []
for point in x:
  model_pred = 0
  #print(point,point.shape)
  for i in range(30):
    point_new = point
    point_new = point_new[list_selected_columns[i]]
    filename = "finalized_model{num}.sav".format(num=i)
    loaded_model = pickle.load(open(filename, 'rb'))
    pred = loaded_model.predict(point_new.reshape(1,-1))
    #print(pred)
    #print(pred)
    model_pred += pred

  all_pred.append(float(model_pred/30))

print(len(all_pred))

506


In [111]:
#MSE
mse = mean_squared_error(y,all_pred)
print('Th MSE is:',mse)

Th MSE is: 2.104805679490111


<font color='orange'><b>Flowchart for calculating MSE </b></font>

![alt text](https://i.imgur.com/sPEE618.png)

After getting predicted_y for each data point, we can use sklearns mean_squared_error to calculate the MSE between predicted_y and actual_y.

*  <font color='blue'><b> Write code for calculating MSE</b></font>

<font color='blue'><b>Step - 3 </b></font>

<font color='orange'><b>Flowchart for calculating OOB score</b></font>

![alt text](https://i.imgur.com/95S5Mtm.png)

Now calculate the $OOB Score =  \frac{1}{506}\sum_{i=1}^{506}(y^{i} - y^{i}_{pred})^{2}$.

*  <font color='blue'><b> Write code for calculating OOB score </b></font>

In [136]:
all_pred = []
for j,point in enumerate(x):
  model_pred = []
  #print(point,point.shape)
  for i in range(30):
    if j  not in list_selected_row[i]:
      point_new = point
      point_new = point_new[list_selected_columns[i]]
      filename = "finalized_model{num}.sav".format(num=i)
      loaded_model = pickle.load(open(filename, 'rb'))
      pred = loaded_model.predict(point_new.reshape(1,-1))
      #print(pred)
      #print(pred)
    model_pred.append(pred)

  all_pred.append(np.median(model_pred))

print(len(all_pred))

506


In [138]:
#OOB
oob = mean_squared_error(y,all_pred)
print('Th OOB is:',oob)

Th OOB is: 12.18094733201581


Observation:

 - We have implemented bootstrap aggregation.
In this we build high variance DT and then combine them parallely to reduce the variance. We also  reduce overfitting which DT are prone to.

- First we created 30 samples from the given dataset.
 We did this by taking 60 percent random points from the dataset then utilizing these point to get more 40 points.
 After this there are point which might occur 2 times in datset.

- After making function for splitting data. We used the function to make 30 samples.

- For each data point we calculated the mse and oob score by aggregating the results obtained from the 30 DT models which we fitted with the data obtained after 30 random split.

- Observed values:

 - The observed mse score is 2.1 
 - The observed oob score is 12.2

- Both of them are less. 




# <font color='blue'><b>Task 2</b></font>

In [ ]:
boston = load_boston()
x=boston.data #independent variables
y=boston.target #target variable

In [149]:
print(y.shape)

(506,)

In [164]:
#Repeating Task for 35


mse_35 = []
oob_35 = []

for task in range(35):
 
  # Creating new datasets
  list_input_data =[]
  list_output_data =[]
  list_selected_row= []
  list_selected_columns=[]

  

  for i in range(0,30):
    a,b,c,d = generating_samples(x,y)
    list_input_data.append(a)
    list_output_data.append(b)
    list_selected_row.append(c)
    list_selected_columns.append(d)

  #print(y.shape)

  list_of_all_models = []

  for i in range(30):
    model = DecisionTreeRegressor(max_depth=None)
    model.fit(list_input_data[i],list_output_data[i])
    filename = "finalized_model{task}{num}.sav".format(task = task,num=i)
    pickle.dump(model, open(filename, 'wb'))
    list_of_all_models.append(filename)
    
  #print(y.shape)

  #Calculating mse
  all_pred = []

  for point in x:
    model_pred = 0
  #print(point,point.shape)
    for i in range(30):
      point_new = point
      point_new = point_new[list_selected_columns[i]]
      filename = "finalized_model{task}{num}.sav".format(task = task,num=i)

      loaded_model = pickle.load(open(filename, 'rb'))
      pred = loaded_model.predict(point_new.reshape(1,-1))
    
      model_pred += pred


    all_pred.append(float(model_pred/30))


  #print(y.shape,'y')
  #print(len(all_pred),'allpred')
  mse_35.append(mean_squared_error(y,all_pred))

  #calculatin oob
  all_pred = []
  for j,point in enumerate(x):
    model_pred = []
    #print(point,point.shape)
    for i in range(30):
      if j  not in list_selected_row[i]:
        point_new = point
        point_new = point_new[list_selected_columns[i]]
        filename = "finalized_model{task}{num}.sav".format(task = task,num=i)

        loaded_model = pickle.load(open(filename, 'rb'))
        pred = loaded_model.predict(point_new.reshape(1,-1))
        #print(pred)
        #print(pred)
      model_pred.append(pred)

    all_pred.append(np.median(model_pred))


  oob_35.append(mean_squared_error(y,all_pred))

  



print(len(oob_35))
print(len(mse_35))




35
35


In [197]:
#Calculatin CI
#https://stackoverflow.com/questions/15033511/compute-a-confidence-interval-from-sample-data
import numpy as np
import scipy.stats


def mean_confidence_interval(data, confidence=0.95):
    a = 1.0 * np.array(data)
    n = len(a)
    m, se = np.mean(a), scipy.stats.sem(a)
    h = se * scipy.stats.t.ppf((1 + confidence) / 2., n-1)
    return m, m-h, m+h

In [200]:
oob_ci = mean_confidence_interval(oob_35)
mse_ci = mean_confidence_interval(mse_35)

print('Mean:','Lower Limit:','Upper_limit:')
print(mse_ci)
print(oob_ci)


Mean: Lower Limit: Upper_limit:
(2.34044610425264, 2.2489379443822677, 2.431954264123012)
(15.71835651531547, 14.960214737658266, 16.476498292972675)


Observation:

 - In this task we calculated the sample using the 35 random data sample using the function of random sampling then calculted mse and oob scores.

 - Using the score as samples we calculated the 95 % confidence interval for mean of the samples. It means if we repeat the experiment then there is 95% chance that the mean will be observed in the given range.
 

 - For MSE 

  - Mean: 2.3, Upper Range: 2.24, Lowe Range:2.4

- For OOB Score

  - Mean: 15.7 Upper Range: 14.96 Lower Range: 16.47




# <font color='blue'><b>Task 3</b></font>

<font color='orange'><b>Flowchart for Task 3</b></font>

<b>Hint: </b> We created 30 models by using 30 samples in TASK-1. Here, we need send query point "xq"  to 30 models and perform the regression on the output generated by 30 models.

![alt text](https://i.imgur.com/Y5cNhQk.png)

*  <font color='blue'><b> Write code for TASK 3 </b></font>

In [185]:
xq = np.array([0.18,20.0,5.00,0.0,0.421,5.60,72.2,7.95,7.0,30.0,19.1,372.13,18.60])
len(xq)

13

In [188]:
xq[list_selected_columns[i]].shape

(5,)

In [178]:
list_of_all_models = []

for i in range(30):

  model = DecisionTreeRegressor(max_depth=None)
  model.fit(list_input_data[i],list_output_data[i])
  filename = "finalized_model_part_3{num}.sav".format(num=i)
  pickle.dump(model, open(filename, 'wb'))
  list_of_all_models.append(filename)


print(len(list_of_all_models))

30


In [196]:
def predict(point):
  model_pred = 0
  for i in range(30):
    #print(i)
    point_new = point
    point_new = point_new[list_selected_columns[i]]

  
    filename = "finalized_model_part_3{num}.sav".format(num=i)

    loaded_model = pickle.load(open(filename, 'rb'))
    #print(point_new.shape)
    #print(loaded_model.get_params())
    #print(point_new.shape,'last')
  
    pred = loaded_model.predict(point_new.reshape(1,-1))
    
    model_pred += pred

  return model_pred/30

prediction = predict(xq)
print('prediction for xq is:',prediction)

prediction for xq is: [21.62166667]


Observation:

 - In this part we implemented predict function. After getting we need to get the columns of point on whihc the DT was trained and get the output.

- After getting the output from all 30 model we retuurn the average as the prediction.

<br><br><br>

<font color='red'><b>Write observations for task 1, task 2, task 3 indetail</b></font>